parsers/CA_NS.py

```diff
@@ -1,212 +1,166 @@
-            }
+        production_breakdowns.append(
+            datetime=timestamp,
+            production=production_mix,
+            source=SOURCE,
+            zoneKey=ZONE_KEY,
         )
 
-    return production, imports
-
-
-def fetch_production(
-    zone_key: str = "CA-NS",
-    session: Session | None = None,
-    target_datetime: datetime | None = None,
-    logger: Logger = getLogger(__name__),
-) -> list[dict]:
-    """Requests the last known production mix (in MW) of a given country."""
-    if target_datetime:
-        raise NotImplementedError(
-            "This parser is unable to give information more than 24 hours in the past"
-        )
-
-    r = session or Session()
-
-    production, imports = _get_ns_info(r, logger)
-
-    return production
+    return production_breakdowns.to_list()
 
 
 def fetch_exchange(
-    zone_key1: str,
-    zone_key2: str,
+    zone_key1: ZoneKey,
+    zone_key2: ZoneKey,
     session: Session | None = None,
     target_datetime: datetime | None = None,
     logger: Logger = getLogger(__name__),
-) -> list[dict]:
+) -> list[dict[str, Any]]:
     """
     Requests the last known power exchange (in MW) between two regions.
-
-    Note: As of early 2017, Nova Scotia only has an exchange with New Brunswick (CA-NB).
-    (An exchange with Newfoundland, "Maritime Link", is scheduled to open in "late 2017").
-
-    The API for Nova Scotia only specifies imports.
-    When NS is exporting energy, the API returns 0.
     """
     if target_datetime:
-        raise NotImplementedError(
-            "This parser is unable to give information more than 24 hours in the past"
-        )
-
-    sorted_zone_keys = "->".join(sorted([zone_key1, zone_key2]))
-
-    if sorted_zone_keys != "CA-NB->CA-NS":
-        raise NotImplementedError("This exchange pair is not implemented")
-
-    requests_obj = session or Session()
-    _, imports = _get_ns_info(requests_obj, logger)
-
-    return imports
+        raise ParserException(PARSER, "Unable to fetch historical data", ZONE_KEY)
+
+    sorted_zone_keys = ZoneKey("->".join(sorted((zone_key1, zone_key2))))
+    if sorted_zone_keys not in (ZoneKey("CA-NB->CA-NS"), ZoneKey("CA-NL-NF->CA-NS")):
+        raise ParserException(PARSER, "Unimplemented exchange pair", sorted_zone_keys)
+
+    session = session or Session()
+    soup = BeautifulSoup(session.get(EXCHANGE_URL).text, "html.parser")
+
+    # Extract the timestamp from the table header.
+    timestamp = datetime.strptime(
+        soup.find(string="Current System Conditions").find_next("td").em.i.string,
+        "%d-%b-%y %H:%M:%S",
+    ).replace(tzinfo=ZoneInfo("America/Halifax"))
+
+    # Choose the appropriate exchange figure for the requested zone pair.
+    exchange = (
+        -float(soup.find(string="NS Export ").find_next("td").string)
+        if sorted_zone_keys == ZoneKey("CA-NB->CA-NS")
+        else float(soup.find(string="Maritime Link Import ").find_next("td").string)
```

Comment:
We need to ensure all these soup.find() are not none before we pass them to functions that can't handle None values for type safety.

Then I think we can merge this.

Reply:
I'm not sure how we should handle this, because it essentially boils down to how the HTML is structured. We could catch errors and set `exchange = None`, for example, but if Nova Scotia Power update the HTML, we might wind up ignoring errors that actually require a parser fix. I think it's reasonable to assume the HTML has a particular structure, and if that changes, Python will let us know by raising an exception.

Reply:
What we should do is add clear error handling and throw exceptions with clear messages if that is the case. Not rely on the whole thing crashing due to the wrong type being used.

If we add type checking to the CI, which we want to do at some point, we'd have to change this anyway.